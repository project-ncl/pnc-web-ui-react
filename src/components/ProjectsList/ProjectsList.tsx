import '@patternfly/react-core/dist/styles/base.css';

import { Table, TableHeader, TableBody, cellWidth } from '@patternfly/react-table';
import { Project } from 'pnc-api-types-ts';

import { Link } from 'react-router-dom';

interface IProjectsList {
  projects: Project[];
}

/**
 * Dump component displaying list of Projects.
 *
 * @param projects - List of Projects to be displayed
 */
export const ProjectsList = ({ projects }: IProjectsList) => {
  const columns = [
    { title: 'Name', transforms: [cellWidth(30)] },
    'Description',
    { title: 'Build Configs count', transforms: [cellWidth(15)] },
  ];

  const rows = projects.map((project: Project) => [
    {
      title: <Link to="#">{project.name}</Link>,
    },
    project.description,
    Object.keys(project.buildConfigs || []).length,
  ]);

  return (
    <Table aria-label="Projects List" variant="compact" borders={false} cells={columns} rows={rows}>
      <TableHeader />
      <TableBody />
    </Table>
  );
};
