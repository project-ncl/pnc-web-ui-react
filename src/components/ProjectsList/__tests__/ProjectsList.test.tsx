import { ProjectsList } from '../ProjectsList';
import { MemoryRouter } from 'react-router-dom';
import { render, screen } from '@testing-library/react';

jest.mock('../../../services/projectService');

describe('display ProjectList component', () => {
  let mockProjectsRequest: any;
  let mockProjects: any;

  async function loadMocks() {
    mockProjectsRequest = await import('../../../services/__mocks__/projects-mock.json');
    mockProjects = mockProjectsRequest.content;
  }

  beforeEach(async () => {
    await loadMocks();
  });

  test('renders ProjectList to have the right data', () => {
    render(
      <MemoryRouter>
        <ProjectsList projects={mockProjects} />
      </MemoryRouter>
    );
    const firstProject = screen.getByText(mockProjects[0].name);
    expect(firstProject).toBeInTheDocument();
    const lastProject = screen.getByText(mockProjects[6].description);
    expect(lastProject).toBeInTheDocument();
    const lastProjectCount = screen.getByText(17);
    expect(lastProjectCount).toBeInTheDocument();
  });

  test('compare snapshot with previous record', () => {
    const tree = render(
      <MemoryRouter>
        <ProjectsList projects={mockProjects} />
      </MemoryRouter>
    );
    expect(tree).toMatchSnapshot();
  });
});
