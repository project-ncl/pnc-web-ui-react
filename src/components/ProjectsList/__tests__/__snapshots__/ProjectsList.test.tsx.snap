// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`display ProjectList component compare snapshot with previous record 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <table
        aria-label="Projects List"
        class="pf-c-table pf-m-grid-md pf-m-compact pf-m-no-border-rows"
        data-ouia-component-id="OUIA-Generated-Table-4"
        data-ouia-component-type="PF4/Table"
        data-ouia-safe="true"
        role="grid"
      >
        <thead
          class=""
        >
          <tr
            class=""
            data-ouia-component-id="OUIA-Generated-TableRow-9"
            data-ouia-component-type="PF4/TableRow"
            data-ouia-safe="true"
          >
            <th
              class="pf-m-width-30"
              data-key="0"
              data-label="Name"
              scope="col"
            >
              Name
            </th>
            <th
              class=""
              data-key="1"
              data-label="Description"
              scope="col"
            >
              Description
            </th>
            <th
              class="pf-m-width-15"
              data-key="2"
              data-label="Build Configs count"
              scope="col"
            >
              Build Configs count
            </th>
            <th
              class=""
              data-key="3"
              data-label="Actions"
              scope="col"
            >
              Actions
            </th>
          </tr>
        </thead>
        <tbody
          class=""
          role="rowgroup"
        >
          <tr
            class=""
            data-ouia-component-id="OUIA-Generated-TableRow-10"
            data-ouia-component-type="PF4/TableRow"
            data-ouia-safe="true"
          >
            <td
              class=""
              data-key="0"
              data-label="Name"
            >
              <a
                href="/100"
              >
                Test Project Name 100
              </a>
            </td>
            <td
              class=""
              data-key="1"
              data-label="Description"
            >
              Test Project Name 100 Desc
            </td>
            <td
              class=""
              data-key="2"
              data-label="Build Configs count"
            >
              0
            </td>
            <td
              class=""
              data-key="3"
              data-label="Actions"
            >
              <a
                href="/100/edit"
              >
                edit
              </a>
            </td>
          </tr>
          <tr
            class=""
            data-ouia-component-id="OUIA-Generated-TableRow-11"
            data-ouia-component-type="PF4/TableRow"
            data-ouia-safe="true"
          >
            <td
              class=""
              data-key="0"
              data-label="Name"
            >
              <a
                href="/101"
              >
                zz_2021110-101
              </a>
            </td>
            <td
              class=""
              data-key="1"
              data-label="Description"
            >
              zz_2021110-101 desc
            </td>
            <td
              class=""
              data-key="2"
              data-label="Build Configs count"
            >
              0
            </td>
            <td
              class=""
              data-key="3"
              data-label="Actions"
            >
              <a
                href="/101/edit"
              >
                edit
              </a>
            </td>
          </tr>
          <tr
            class=""
            data-ouia-component-id="OUIA-Generated-TableRow-12"
            data-ouia-component-type="PF4/TableRow"
            data-ouia-safe="true"
          >
            <td
              class=""
              data-key="0"
              data-label="Name"
            >
              <a
                href="/102"
              >
                zz_2021110-102
              </a>
            </td>
            <td
              class=""
              data-key="1"
              data-label="Description"
            />
            <td
              class=""
              data-key="2"
              data-label="Build Configs count"
            >
              0
            </td>
            <td
              class=""
              data-key="3"
              data-label="Actions"
            >
              <a
                href="/102/edit"
              >
                edit
              </a>
            </td>
          </tr>
          <tr
            class=""
            data-ouia-component-id="OUIA-Generated-TableRow-13"
            data-ouia-component-type="PF4/TableRow"
            data-ouia-safe="true"
          >
            <td
              class=""
              data-key="0"
              data-label="Name"
            >
              <a
                href="/103"
              >
                zz_2021110-103
              </a>
            </td>
            <td
              class=""
              data-key="1"
              data-label="Description"
            />
            <td
              class=""
              data-key="2"
              data-label="Build Configs count"
            >
              0
            </td>
            <td
              class=""
              data-key="3"
              data-label="Actions"
            >
              <a
                href="/103/edit"
              >
                edit
              </a>
            </td>
          </tr>
          <tr
            class=""
            data-ouia-component-id="OUIA-Generated-TableRow-14"
            data-ouia-component-type="PF4/TableRow"
            data-ouia-safe="true"
          >
            <td
              class=""
              data-key="0"
              data-label="Name"
            >
              <a
                href="/104"
              >
                zz_2021110-104
              </a>
            </td>
            <td
              class=""
              data-key="1"
              data-label="Description"
            />
            <td
              class=""
              data-key="2"
              data-label="Build Configs count"
            >
              0
            </td>
            <td
              class=""
              data-key="3"
              data-label="Actions"
            >
              <a
                href="/104/edit"
              >
                edit
              </a>
            </td>
          </tr>
          <tr
            class=""
            data-ouia-component-id="OUIA-Generated-TableRow-15"
            data-ouia-component-type="PF4/TableRow"
            data-ouia-safe="true"
          >
            <td
              class=""
              data-key="0"
              data-label="Name"
            >
              <a
                href="/105"
              >
                zz_2021110-105
              </a>
            </td>
            <td
              class=""
              data-key="1"
              data-label="Description"
            >
              zz_2021110-105 description
            </td>
            <td
              class=""
              data-key="2"
              data-label="Build Configs count"
            >
              0
            </td>
            <td
              class=""
              data-key="3"
              data-label="Actions"
            >
              <a
                href="/105/edit"
              >
                edit
              </a>
            </td>
          </tr>
          <tr
            class=""
            data-ouia-component-id="OUIA-Generated-TableRow-16"
            data-ouia-component-type="PF4/TableRow"
            data-ouia-safe="true"
          >
            <td
              class=""
              data-key="0"
              data-label="Name"
            >
              <a
                href="/106"
              >
                Test Project Name 106
              </a>
            </td>
            <td
              class=""
              data-key="1"
              data-label="Description"
            >
              Test Project Name 106 description
            </td>
            <td
              class=""
              data-key="2"
              data-label="Build Configs count"
            >
              17
            </td>
            <td
              class=""
              data-key="3"
              data-label="Actions"
            >
              <a
                href="/106/edit"
              >
                edit
              </a>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </body>,
  "container": <div>
    <table
      aria-label="Projects List"
      class="pf-c-table pf-m-grid-md pf-m-compact pf-m-no-border-rows"
      data-ouia-component-id="OUIA-Generated-Table-4"
      data-ouia-component-type="PF4/Table"
      data-ouia-safe="true"
      role="grid"
    >
      <thead
        class=""
      >
        <tr
          class=""
          data-ouia-component-id="OUIA-Generated-TableRow-9"
          data-ouia-component-type="PF4/TableRow"
          data-ouia-safe="true"
        >
          <th
            class="pf-m-width-30"
            data-key="0"
            data-label="Name"
            scope="col"
          >
            Name
          </th>
          <th
            class=""
            data-key="1"
            data-label="Description"
            scope="col"
          >
            Description
          </th>
          <th
            class="pf-m-width-15"
            data-key="2"
            data-label="Build Configs count"
            scope="col"
          >
            Build Configs count
          </th>
          <th
            class=""
            data-key="3"
            data-label="Actions"
            scope="col"
          >
            Actions
          </th>
        </tr>
      </thead>
      <tbody
        class=""
        role="rowgroup"
      >
        <tr
          class=""
          data-ouia-component-id="OUIA-Generated-TableRow-10"
          data-ouia-component-type="PF4/TableRow"
          data-ouia-safe="true"
        >
          <td
            class=""
            data-key="0"
            data-label="Name"
          >
            <a
              href="/100"
            >
              Test Project Name 100
            </a>
          </td>
          <td
            class=""
            data-key="1"
            data-label="Description"
          >
            Test Project Name 100 Desc
          </td>
          <td
            class=""
            data-key="2"
            data-label="Build Configs count"
          >
            0
          </td>
          <td
            class=""
            data-key="3"
            data-label="Actions"
          >
            <a
              href="/100/edit"
            >
              edit
            </a>
          </td>
        </tr>
        <tr
          class=""
          data-ouia-component-id="OUIA-Generated-TableRow-11"
          data-ouia-component-type="PF4/TableRow"
          data-ouia-safe="true"
        >
          <td
            class=""
            data-key="0"
            data-label="Name"
          >
            <a
              href="/101"
            >
              zz_2021110-101
            </a>
          </td>
          <td
            class=""
            data-key="1"
            data-label="Description"
          >
            zz_2021110-101 desc
          </td>
          <td
            class=""
            data-key="2"
            data-label="Build Configs count"
          >
            0
          </td>
          <td
            class=""
            data-key="3"
            data-label="Actions"
          >
            <a
              href="/101/edit"
            >
              edit
            </a>
          </td>
        </tr>
        <tr
          class=""
          data-ouia-component-id="OUIA-Generated-TableRow-12"
          data-ouia-component-type="PF4/TableRow"
          data-ouia-safe="true"
        >
          <td
            class=""
            data-key="0"
            data-label="Name"
          >
            <a
              href="/102"
            >
              zz_2021110-102
            </a>
          </td>
          <td
            class=""
            data-key="1"
            data-label="Description"
          />
          <td
            class=""
            data-key="2"
            data-label="Build Configs count"
          >
            0
          </td>
          <td
            class=""
            data-key="3"
            data-label="Actions"
          >
            <a
              href="/102/edit"
            >
              edit
            </a>
          </td>
        </tr>
        <tr
          class=""
          data-ouia-component-id="OUIA-Generated-TableRow-13"
          data-ouia-component-type="PF4/TableRow"
          data-ouia-safe="true"
        >
          <td
            class=""
            data-key="0"
            data-label="Name"
          >
            <a
              href="/103"
            >
              zz_2021110-103
            </a>
          </td>
          <td
            class=""
            data-key="1"
            data-label="Description"
          />
          <td
            class=""
            data-key="2"
            data-label="Build Configs count"
          >
            0
          </td>
          <td
            class=""
            data-key="3"
            data-label="Actions"
          >
            <a
              href="/103/edit"
            >
              edit
            </a>
          </td>
        </tr>
        <tr
          class=""
          data-ouia-component-id="OUIA-Generated-TableRow-14"
          data-ouia-component-type="PF4/TableRow"
          data-ouia-safe="true"
        >
          <td
            class=""
            data-key="0"
            data-label="Name"
          >
            <a
              href="/104"
            >
              zz_2021110-104
            </a>
          </td>
          <td
            class=""
            data-key="1"
            data-label="Description"
          />
          <td
            class=""
            data-key="2"
            data-label="Build Configs count"
          >
            0
          </td>
          <td
            class=""
            data-key="3"
            data-label="Actions"
          >
            <a
              href="/104/edit"
            >
              edit
            </a>
          </td>
        </tr>
        <tr
          class=""
          data-ouia-component-id="OUIA-Generated-TableRow-15"
          data-ouia-component-type="PF4/TableRow"
          data-ouia-safe="true"
        >
          <td
            class=""
            data-key="0"
            data-label="Name"
          >
            <a
              href="/105"
            >
              zz_2021110-105
            </a>
          </td>
          <td
            class=""
            data-key="1"
            data-label="Description"
          >
            zz_2021110-105 description
          </td>
          <td
            class=""
            data-key="2"
            data-label="Build Configs count"
          >
            0
          </td>
          <td
            class=""
            data-key="3"
            data-label="Actions"
          >
            <a
              href="/105/edit"
            >
              edit
            </a>
          </td>
        </tr>
        <tr
          class=""
          data-ouia-component-id="OUIA-Generated-TableRow-16"
          data-ouia-component-type="PF4/TableRow"
          data-ouia-safe="true"
        >
          <td
            class=""
            data-key="0"
            data-label="Name"
          >
            <a
              href="/106"
            >
              Test Project Name 106
            </a>
          </td>
          <td
            class=""
            data-key="1"
            data-label="Description"
          >
            Test Project Name 106 description
          </td>
          <td
            class=""
            data-key="2"
            data-label="Build Configs count"
          >
            17
          </td>
          <td
            class=""
            data-key="3"
            data-label="Actions"
          >
            <a
              href="/106/edit"
            >
              edit
            </a>
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
